File: codeql-custom-queries-java/lang/ManualArrayCopy.java
public class ManualArrayCopy {

    public void testManualCopies() {
        int[] src = {1, 2, 3, 4, 5};
        int[] dest = new int[5];

        for (int i = 0; i < src.length; i++) {
            dest[i] = src[i]; 
        }

        int j = 0;
        while (j < src.length) {
            dest[j] = src[j];
            j++;
        }

        int index = 0;
        for (int value : src) {
            dest[index++] = value;
        }

        if (dest.length > 0) {
            for (int k = 0; k < src.length; k++) {
                dest[k] = src[k];
            }
        }
    }

    public void testValidCases() {
        int[] src = {1, 2, 3, 4, 5};
        int[] dest = new int[5];

        System.arraycopy(src, 0, dest, 0, src.length);

        for (int i = 0; i < src.length; i++) {
            src[i] = src[i] * 2;
        }

        for (int i = 0; i < dest.length; i++) {
            dest[i] = 0;
        }
    }
}

File: codeql-custom-queries-java/lang/ManualArrayCopy.kt
class ManualArrayCopy {

    // Devrait déclencher 4 à 5 alertes selon la précision de la requête
    fun testManualCopies() {
        val src = intArrayOf(1, 2, 3, 4, 5)
        val dest = IntArray(5)

        // 1. Boucle for classique (indexée)
        for (i in src.indices) {
            dest[i] = src[i]
        }

        // 2. Boucle while
        var j = 0
        while (j < src.size) {
            dest[j] = src[j]
            j++
        }

        // 3. Boucle for-each (Enhanced for)
        var index = 0
        for (value in src) {
            dest[index++] = value
        }

        // 4. Boucle imbriquée dans une condition
        if (dest.isNotEmpty()) {
            for (k in src.indices) {
                dest[k] = src[k]
            }
        }
    }

    // Ne devrait pas déclencher d'alerte
    fun testValidCases() {
        val src = intArrayOf(1, 2, 3, 4, 5)
        val dest = IntArray(5)

        // La manière correcte en Kotlin (équivalent System.arraycopy)
        src.copyInto(dest)

        // Manipulation de données (pas une simple copie)
        for (i in src.indices) {
            src[i] = src[i] * 2
        }

        // Initialisation (pas une copie d'un autre tableau)
        for (i in dest.indices) {
            dest[i] = 0
        }
    }
}